extensions/guacamole-vault/modules/guacamole-vault-ksm/src/main/java/org/apache/guacamole/vault/ksm/user/KsmDirectoryService.java

```diff
@@ -52,218 +39,160 @@
+                        connectionGroup.getAttributes()));
 
-                // Create a map to store the extracted values from the KSM storage
-                Map<String, String> configMap = new HashMap<>();
+                super.add(connectionGroup);
+            }
 
-                // Go through all the expected fields, extract from the KSM storage,
-                // and write to the newly created map
-                EXPECTED_KSM_FIELDS.forEach(configKey -> {
+            @Override
+            public void update(ConnectionGroup connectionGroup) throws GuacamoleException {
 
-                    // Only write the value into the new map if non-null
-                    String value = storage.getString(configKey);
-                    if (value != null)
-                        configMap.put(configKey, value);
+                // Unwrap the existing ConnectionGroup
+                if (connectionGroup instanceof KsmConnectionGroup)
+                    connectionGroup = ((KsmConnectionGroup) connectionGroup).getUnderlyingConnectionGroup();
 
-                });
+                // Process attribute values before saving
+                connectionGroup.setAttributes(
+                    attributeService.processAttributes(
+                        connectionGroup.getAttributes()));
 
-                // JSON-encode the value, and then base64 encode that to get the format
-                // that KSM would expect
-                String jsonString = objectMapper.writeValueAsString(configMap);
-                String base64EncodedJson = Base64.getEncoder().encodeToString(
-                        jsonString.getBytes(StandardCharsets.UTF_8));
+                super.update(connectionGroup);
+
+            }
 
-                // Finally, try to parse the newly generated token as a KSM config. If this
-                // works, the config should be fully functional
-                KsmConfig.parseKsmConfig(base64EncodedJson);
+            @Override
+            public ConnectionGroup get(String identifier) throws GuacamoleException {
 
-                // Make a copy of the existing attributes, modifying just the value for
-                // KSM_CONFIGURATION_ATTRIBUTE
-                attributes = new HashMap<>(attributes);
-                attributes.put(
-                        KsmAttributeService.KSM_CONFIGURATION_ATTRIBUTE, base64EncodedJson);
+                // Fetch the ConnectionGroup
+                ConnectionGroup connectionGroup = super.get(identifier);
 
-                // Set the newly updated attributes back to the original object
-                entity.setAttributes(attributes);
+                // Do not process the ConnectionGroup further if it does not exist
+                if (connectionGroup == null)
+                    return null;
 
-            }
+                // Unwrap the existing ConnectionGroup before rewrapping
+                if (connectionGroup instanceof KsmConnectionGroup)
```

Comment:
I'm a bit confused by this. Under what circumstance would `super.get()` return a `KsmConnectionGroup`?

Reply:
Looks like this is just copypasta. I'll remove.